File: ios/Classes/ErikPlatformView.swift
import Flutter
import UIKit

final class ErikPlatformView: NSObject, FlutterPlatformView {
    private let channel: FlutterMethodChannel
    private let containerView: UIView
    private let erikViewController: ErikViewController?

    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        assetResolver: ErikAssetResolver
    ) {
        self.channel = FlutterMethodChannel(
            name: "erik_flutter_sdk/view_\(viewId)",
            binaryMessenger: messenger
        )
        self.containerView = UIView(frame: frame)

        do {
            let erikViewController = try ErikViewController.makeWithBundleAssets(resolver: assetResolver)
            self.erikViewController = erikViewController
            super.init()
            channel.setMethodCallHandler(handle)
            erikViewController.onStateChanged = { [weak self] state in
                self?.channel.invokeMethod("onStateChanged", arguments: state.toMap())
            }
            attach(viewController: erikViewController)
            channel.invokeMethod("onStateChanged", arguments: erikViewController.currentState().toMap())
        } catch {
            self.erikViewController = nil
            super.init()
            channel.setMethodCallHandler(handle)
            let label = UILabel(frame: frame)
            label.numberOfLines = 0
            label.textAlignment = .center
            label.textColor = .white
            label.text = "ErikView failed to load iOS assets.\n\(error.localizedDescription)"
            label.backgroundColor = .black
            label.autoresizingMask = [.flexibleWidth, .flexibleHeight]
            containerView.backgroundColor = .black
            containerView.addSubview(label)
        }
    }

    deinit {
        channel.setMethodCallHandler(nil)
        erikViewController?.dispose()
    }

    func view() -> UIView {
        containerView
    }

    private func attach(viewController: ErikViewController) {
        viewController.loadViewIfNeeded()
        let hostedView = viewController.view!
        hostedView.translatesAutoresizingMaskIntoConstraints = false
        containerView.backgroundColor = .black
        containerView.addSubview(hostedView)
        NSLayoutConstraint.activate([
            hostedView.topAnchor.constraint(equalTo: containerView.topAnchor),
            hostedView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            hostedView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            hostedView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
        ])
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let erikViewController else {
            result(
                FlutterError(
                    code: "view_unavailable",
                    message: "ErikView is unavailable on iOS.",
                    details: nil
                )
            )
            return
        }

        switch call.method {
        case "getState":
            result(erikViewController.currentState().toMap())
        case "goExterior":
            erikViewController.goExterior(completion: commandResult(result))
        case "goInterior":
            erikViewController.goInterior(completion: commandResult(result))
        case "toggleLights":
            erikViewController.toggleLights(completion: commandResult(result))
        case "setColor":
            guard let colorName = Self.colorName(from: call.arguments) else {
                result(
                    FlutterError(
                        code: "invalid_arguments",
                        message: "Expected a non-empty color name.",
                        details: nil
                    )
                )
                return
            }
            erikViewController.setColor(colorName, completion: commandResult(result))
        case "skipIntro":
            erikViewController.skipIntro(completion: commandResult(result))
        case "openDoor":
            guard let door = Self.door(from: call.arguments) else {
                result(
                    FlutterError(
                        code: "invalid_arguments",
                        message: "Expected a valid door name.",
                        details: nil
                    )
                )
                return
            }
            erikViewController.openDoor(door, completion: commandResult(result))
        case "closeDoor":
            guard let door = Self.door(from: call.arguments) else {
                result(
                    FlutterError(
                        code: "invalid_arguments",
                        message: "Expected a valid door name.",
                        details: nil
                    )
                )
                return
            }
            erikViewController.closeDoor(door, completion: commandResult(result))
        case "setAllDoorsOpen":
            guard let open = Self.openFlag(from: call.arguments) else {
                result(
                    FlutterError(
                        code: "invalid_arguments",
                        message: "Expected a boolean open flag.",
                        details: nil
                    )
                )
                return
            }
            erikViewController.setAllDoorsOpen(open, completion: commandResult(result))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func commandResult(_ result: @escaping FlutterResult) -> (Result<Void, Error>) -> Void {
        { commandResult in
            switch commandResult {
            case .success:
                result(nil)
            case let .failure(error):
                result(
                    FlutterError(
                        code: "erik_command_failed",
                        message: error.localizedDescription,
                        details: nil
                    )
                )
            }
        }
    }

    private static func colorName(from arguments: Any?) -> String? {
        switch arguments {
        case let colorName as String where !colorName.isEmpty:
            colorName
        case let payload as [AnyHashable: Any]:
            (payload["color"] as? String)?.nilIfEmpty
        default:
            nil
        }
    }

    private static func openFlag(from arguments: Any?) -> Bool? {
        switch arguments {
        case let open as Bool:
            open
        case let payload as [AnyHashable: Any]:
            payload["open"] as? Bool
        default:
            nil
        }
    }

    private static func door(from arguments: Any?) -> ErikDoor? {
        let rawDoor: String?
        switch arguments {
        case let door as String:
            rawDoor = door
        case let payload as [AnyHashable: Any]:
            rawDoor = payload["door"] as? String
        default:
            rawDoor = nil
        }

        guard let rawDoor else {
            return nil
        }

        return ErikDoor(rawValue: rawDoor)
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: android/src/main/kotlin/com/eccentric/erik_flutter/ErikPlatformView.kt
package com.eccentric.erik_flutter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView
import androidx.core.view.setPadding
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentContainerView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class ErikPlatformView(
    context: Context,
    messenger: BinaryMessenger,
    private val viewId: Int,
    private val activityProvider: () -> FragmentActivity?,
) : PlatformView,
    MethodChannel.MethodCallHandler,
    ErikFragment.Listener {
    private val channel = MethodChannel(messenger, "$VIEW_CHANNEL_PREFIX$viewId")
    private val containerView: View
    private val fragmentTag = "erik_flutter_sdk_fragment_$viewId"

    private var fragmentActivity: FragmentActivity? = null
    private var erikFragment: ErikFragment? = null
    private var disposed = false

    init {
        channel.setMethodCallHandler(this)
        containerView = attachOrCreateContainer(context)
        dispatchState(erikFragment?.currentState() ?: ErikRuntimeState())
    }

    override fun getView(): View = containerView

    override fun dispose() {
        disposed = true
        erikFragment?.setListener(null)
        channel.setMethodCallHandler(null)

        val activity = fragmentActivity
        val fragment = erikFragment
        if (activity != null && fragment != null) {
            activity.supportFragmentManager
                .beginTransaction()
                .remove(fragment)
                .commitNowAllowingStateLoss()
        }

        erikFragment = null
        fragmentActivity = null
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val fragment =
            erikFragment ?: run {
                result.error(
                    "fragment_unavailable",
                    "ErikFragment is unavailable. Ensure the host Activity extends FlutterFragmentActivity.",
                    null,
                )
                return
            }

        when (call.method) {
            "getState" -> result.success(fragment.currentState().toMap())
            "goExterior" -> fragment.goExterior(result.asCommandCallback())
            "goInterior" -> fragment.goInterior(result.asCommandCallback())
            "toggleLights" -> fragment.toggleLights(result.asCommandCallback())
            "setColor" -> {
                val colorName = call.arguments.asColorName()
                if (colorName == null) {
                    result.error("invalid_arguments", "Expected a non-empty color name.", null)
                } else {
                    fragment.setColor(colorName, result.asCommandCallback())
                }
            }
            "skipIntro" -> fragment.skipIntro(result.asCommandCallback())
            "openDoor" -> {
                val door = call.arguments.asDoor()
                if (door == null) {
                    result.error("invalid_arguments", "Expected a valid door name.", null)
                } else {
                    fragment.openDoor(door, result.asCommandCallback())
                }
            }
            "closeDoor" -> {
                val door = call.arguments.asDoor()
                if (door == null) {
                    result.error("invalid_arguments", "Expected a valid door name.", null)
                } else {
                    fragment.closeDoor(door, result.asCommandCallback())
                }
            }
            "setAllDoorsOpen" -> {
                val open = call.arguments.asOpenFlag()
                if (open == null) {
                    result.error("invalid_arguments", "Expected a boolean open flag.", null)
                } else {
                    fragment.setAllDoorsOpen(open, result.asCommandCallback())
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onStateChanged(state: ErikRuntimeState) {
        dispatchState(state)
    }

    private fun attachOrCreateContainer(context: Context): View {
        val activity = activityProvider()
        if (activity == null) {
            return TextView(context).apply {
                layoutParams =
                    ViewGroup.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT,
                    )
                text =
                    "ErikFragment requires a FragmentActivity host. Use FlutterFragmentActivity in the Flutter example."
                setPadding(32)
            }
        }

        fragmentActivity = activity
        val containerId = VIEW_CONTAINER_BASE_ID + viewId
        val fragmentContainerView =
            FragmentContainerView(context).apply {
                id = containerId
                layoutParams =
                    FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT,
                    )
            }

        val fragmentManager = activity.supportFragmentManager
        val existingFragment = fragmentManager.findFragmentByTag(fragmentTag) as? ErikFragment
        val fragment = existingFragment ?: ErikFragment()
        fragment.setListener(this)
        erikFragment = fragment

        if (existingFragment == null) {
            fragmentContainerView.replaceWithFragmentWhenAttached(containerId, fragment)
        }

        return fragmentContainerView
    }

    private fun FragmentContainerView.replaceWithFragmentWhenAttached(
        containerId: Int,
        fragment: ErikFragment,
    ) {
        if (isAttachedToWindow) {
            replaceWithFragment(containerId, fragment)
            return
        }

        addOnAttachStateChangeListener(
            object : View.OnAttachStateChangeListener {
                override fun onViewAttachedToWindow(view: View) {
                    removeOnAttachStateChangeListener(this)
                    replaceWithFragment(containerId, fragment)
                }

                override fun onViewDetachedFromWindow(view: View) = Unit
            },
        )
    }

    private fun replaceWithFragment(
        containerId: Int,
        fragment: ErikFragment,
    ) {
        val activity = fragmentActivity ?: return
        if (disposed || activity.supportFragmentManager.isDestroyed) {
            return
        }

        activity.supportFragmentManager
            .beginTransaction()
            .replace(containerId, fragment, fragmentTag)
            .commitNowAllowingStateLoss()
    }

    private fun dispatchState(state: ErikRuntimeState) {
        channel.invokeMethod("onStateChanged", state.toMap())
    }

    private fun MethodChannel.Result.asCommandCallback(): ErikCommandCallback =
        object : ErikCommandCallback {
            override fun onSuccess() {
                success(null)
            }

            override fun onError(message: String) {
                error("erik_command_failed", message, null)
            }
        }

    private fun ErikRuntimeState.toMap(): Map<String, Any> =
        mapOf(
            "isReady" to isReady,
            "isIntroAnimationPlaying" to isIntroAnimationPlaying,
        )

    private fun Any?.asColorName(): String? =
        when (this) {
            is String -> takeIf(String::isNotBlank)
            is Map<*, *> -> (this["color"] as? String)?.takeIf(String::isNotBlank)
            else -> null
        }

    private fun Any?.asOpenFlag(): Boolean? =
        when (this) {
            is Boolean -> this
            is Map<*, *> -> this["open"] as? Boolean
            else -> null
        }

    private fun Any?.asDoor(): ErikDoor? {
        val rawDoor =
            when (this) {
                is String -> this
                is Map<*, *> -> this["door"] as? String
                else -> null
            }

        return when (rawDoor) {
            "frontLeft" -> ErikDoor.FRONT_LEFT
            "frontRight" -> ErikDoor.FRONT_RIGHT
            "rearLeft" -> ErikDoor.REAR_LEFT
            "rearRight" -> ErikDoor.REAR_RIGHT
            "boot" -> ErikDoor.BOOT
            "sunroof" -> ErikDoor.SUNROOF
            else -> null
        }
    }

    companion object {
        const val VIEW_TYPE = "erik_flutter_sdk/erik_fragment_view"
        private const val VIEW_CHANNEL_PREFIX = "erik_flutter_sdk/view_"
        private const val VIEW_CONTAINER_BASE_ID = 47000
    }
}
